File: web/packages/contracts/src/interaction.ts
/**
 * @fileoverview 节点联动契约：控件事件 → 显隐/互斥切换/弹窗/跨屏跳转。
 * 运行态与持久态严格分离——联动只改易失的运行时显隐，绝不写回节点配置；
 * 规则存大屏级 `chromeJson.interactions`。
 *
 * ⚠ 节点引用一律用 `nodeId`（画布节点 id，前端创建时生成、永不再变，ADR-0012）。
 * 不引入第二套编辑期键：id 稳定是这套契约的地基，再造一个键只会多一处能错位的地方。
 * 也别与 `nodeKey` 混淆——那是采集点位的身份，不是画布节点。
 */

/** 控件能上抛的事件名。 */
export const INTERACTION_EVENTS = ['click', 'change', 'select'] as const
export type InteractionEventName = (typeof INTERACTION_EVENTS)[number]

/** 控件上抛的一次交互事件。 */
export interface InteractionEvent {
  event: InteractionEventName
  /** 携带值，如分段切换选中的 `item.key`。 */
  value?: unknown
}

/** show/hide/toggle：对一组目标节点改运行时显隐。 */
export interface InteractionShowAction {
  type: 'show' | 'hide' | 'toggle'
  /** 目标节点 id。 */
  targets: string[]
}

/** setActive：互斥切换组——选中值命中的组显示、其余组隐藏。 */
export interface InteractionSetActiveAction {
  type: 'setActive'
  groups: { value: string; targets: string[] }[]
}

/**
 * openModal：把某个节点连同整棵子树浮起为弹窗。
 * 不带任何内容字段——弹窗内容的单一真源永远是画布上那个节点，
 * 不会出现「联动面板里配了一套、画布上摆了另一套」的分裂。
 * 目标节点通常设为初始不可见，否则屏上与弹窗里各画一份。
 */
export interface InteractionOpenModalAction {
  type: 'openModal'
  /** 弹窗内容根节点 id。 */
  target: string
  /** 留空 = 不渲染标题栏，只留关闭按钮。 */
  title?: string
}

/** closeModal：关闭当前弹窗。弹窗自带关闭键/Esc/点遮罩，不配这条也关得掉。 */
export interface InteractionCloseModalAction {
  type: 'closeModal'
}

/**
 * 跳到哪张大屏。**登录态是大屏 id，公开态是目标屏的公开令牌**——
 * 含义由渲染这棵树的宿主决定，联动引擎只搬运不解释。
 *
 * ⚠ 它**不是 URL**：能配任意地址的大屏等于一个站内跳板（开放重定向），
 * 与路由守卫 `safeReturnTarget` 只放行站内相对路径是同一个理由。
 * ⚠ 也不许拿它跟当前大屏 id 比对——那样公开态那条路（句柄是令牌）就再也接不上。
 */
export type DashboardHandle = string

/**
 * navigate：离开本屏，跳到另一张大屏。
 * ⚠ 这是唯一一档**跨文档**的动作：同一条事件上再配显隐类动作没有意义，页面都换了。
 */
export interface InteractionNavigateAction {
  type: 'navigate'
  target: DashboardHandle
}

/**
 * navigateByValue：按控件上抛的值分流跳转（一块卡片六个指标，各进各的明细屏）。
 *
 * 形状照 `setActive` 的 `groups` 抄：那是本仓已有的「按值分派」口径，
 * 再造一套只会让两处的空值语义各漂各的。
 * ⚠ **没带值的事件一律不跳**：整块可点的模块（`hostClickable`）上抛的 click
 * 是没有 value 的，不挡的话它会命中「值留空」的那条路由，表现成随手点一下就换屏。
 * ⚠ 同一个 value 出现两次时取**第一条**命中，后面那条永远不生效——编辑面要能看出重复。
 */
export interface InteractionNavigateByValueAction {
  type: 'navigateByValue'
  routes: { value: string; target: DashboardHandle }[]
}

export type InteractionAction =
  | InteractionShowAction
  | InteractionSetActiveAction
  | InteractionOpenModalAction
  | InteractionCloseModalAction
  | InteractionNavigateAction
  | InteractionNavigateByValueAction

/** 一条联动规则。 */
export interface InteractionRule {
  id: string
  source: {
    /** 事件源节点 id。 */
    nodeId: string
    event: InteractionEventName
  }
  action: InteractionAction
}

File: server/services/ai-assistant/src/ai_assistant/apps/chat/services/tools/providers/client_specs/interaction.py
"""联动那一批客户端工具的规格：读规则表、写一条、删一条、改初始显隐。

⚠ 与同目录的 `core.py` 同一口径——只有形状，实现在浏览器里；名字必须与
前端工作面的 `tools` 数组逐字相同。对不上时模型看得见那个工具、
调用却每次都失败，而失败的样子与「这一页没实现它」一模一样。

单独一个模块而不是接着往那一份里加：规格文件已经四百多行，而工具还会一直加，
挤在一起的那一天最省事的改法是把新工具塞进别的段落里——名字与实现于是开始漂。
"""

from ai_assistant.apps.chat.services.tools.shapes import (
    ToolSpec,
    object_schema,
    string_schema,
)

INTERACTION_SPECS: tuple[ToolSpec, ...] = (
    ToolSpec(
        name="dashboard.read_interactions",
        description=(
            "读这一屏的全部**联动规则**（点一下让别处显示 / 隐藏 / 切换、"
            "弹窗、跨屏跳转），外加「哪些画布节点能当触发源、各自发得出"
            "什么事件」。动手之前先读一次。"
            "⚠ 只有 `sources` 里列出的那几个节点能当触发源。别的节点配了"
            "规则也永远不触发——一块纯装饰的图片不上抛任何事件，"
            "而规则照样存得下去、不报错。"
            "⚠ 每条规则带一格 `problems`：非空就是这条现在已经是哑的"
            "（源节点被删了、源发不出那个事件、目标节点被删了）。"
            "用户说「联动不好使」时先看它，不要急着新加一条。"
        ),
        parameters=object_schema({}, []),
        runs_on="client",
    ),
    ToolSpec(
        name="dashboard.write_interaction",
        description=(
            "写一条联动规则：`rule_id` 给了就改那一条，不给就新建一条。"
            "一条规则 = 「某个画布节点上发生某个事件」→「做一个动作」。"
            "\n`action` 按 `type` 取这几档之一：\n"
            "- `{type:'show'|'hide'|'toggle', targets:[节点id,…]}` "
            "改一组节点的显隐，最常用的一档\n"
            "- `{type:'setActive', groups:[{value, targets:[…]},…]}` "
            "按上抛的值互斥切换：命中的那组显示、其余组隐藏\n"
            "- `{type:'openModal', target:节点id, title?:'标题'}` "
            "把那个节点连同子树浮成弹窗\n"
            "- `{type:'closeModal'}` 关掉当前弹窗\n"
            "- `{type:'navigate', target:大屏id}` 跳到另一张大屏\n"
            "- `{type:'navigateByValue', routes:[{value, target:大屏id},…]}` "
            "按上抛的值分流跳转\n"
            "⚠ 两档跳转的 `target` 是**另一张大屏的 id**，不是画布节点 id——"
            "从 `dashboards.list` 取。填成节点 id 存得下去、也不报错，"
            "点下去只会跳到一张不存在的屏（这一条会当场被拒）。\n"
            "⚠ 弹窗的目标节点要先设成**初始隐藏**（`dashboard.set_visible`），"
            "否则屏上与弹窗里各画一份。\n"
            "⚠ 按值那两档的 `value` 不能留空：不带值的事件一律不派发，"
            "留空的那条永远不命中。\n"
            "⚠ 联动**只改运行时显隐**，不写回节点配置；规则不在撤销栈上，"
            "用户按 Ctrl+Z 退不回来，改完要跟他说一句。"
        ),
        parameters=object_schema(
            {
                "rule_id": string_schema(
                    "要改哪一条；新建就不给。取自 read_interactions"
                ),
                "source_node_id": string_schema(
                    "事件从哪个画布节点发出，须在 read_interactions 的 "
                    "`sources` 里"
                ),
                "event": {
                    "type": "string",
                    "enum": ["click", "change", "select"],
                    "description": (
                        "触发事件；须是源模块发得出的那几个之一"
                        "（见 `sources[].events`）"
                    ),
                },
                "action": {
                    "type": "object",
                    "description": (
                        "做什么，形状按上面列的那几档之一给；`type` 必给"
                    ),
                    "additionalProperties": True,
                },
            },
            ["source_node_id", "event", "action"],
        ),
        runs_on="client",
    ),
    ToolSpec(
        name="dashboard.remove_interaction",
        description=(
            "删掉一条联动规则。"
            "⚠ 规则不在撤销栈上，**删之前先用 `user.ask` 让用户点一下确认**。"
            "回执里把整条原样交还，用户反悔就照它再写一次。"
        ),
        parameters=object_schema(
            {"rule_id": string_schema("要删哪一条，取自 read_interactions")},
            ["rule_id"],
        ),
        runs_on="client",
    ),
    ToolSpec(
        name="dashboard.set_visible",
        description=(
            "改一个画布节点**保存下来的**初始显隐。"
            "配弹窗与显隐类联动要的另一半：先把那个节点设成初始隐藏，"
            "再让规则在运行时把它显示出来。"
            "⚠ 联动改的是运行时的临时显隐，落库的初始值只有这个工具改得动。"
            "⚠ 隐藏一个容器节点会把它整棵子树一起藏掉。"
        ),
        parameters=object_schema(
            {
                "node_id": string_schema("画布节点 id"),
                "is_visible": {
                    "type": "boolean",
                    "description": "显示给 true，初始隐藏给 false",
                },
            },
            ["node_id", "is_visible"],
        ),
        runs_on="client",
    ),
)
